dockstore-client/src/main/java/io/dockstore/client/cli/nested/WesLauncher.java

```diff
@@ -0,0 +1,200 @@
+        return workflowClient.getWorkflowsApi().getPublishedWorkflowByPath(path, null, null, version);
+    }
+
+    /**
+     * Creates a TRS url for a Dockstore entry
+     *
+     * @param workflowClient The WorkflowClient
+     * @param workflowEntry The workflow entry (i.e. github.com/myRepo/myWorkflow:version)
+     * @param workflow The workflow object we are creating a TRS url for
+     * @return A string representing a TRS URL for an entry
+     */
+    public static String combineTrsUrlComponents(WorkflowClient workflowClient, String workflowEntry, Workflow workflow) {
+        ApiClient client = workflowClient.getClient().getGa4Ghv20Api().getApiClient();
+
+        // Entries are passed in the form {PATH}:{VERSION} or {PATH}
+        final String[] pathAndVersion = workflowEntry.split(":");
+        final String path = pathAndVersion[0];
+
+        // Calculate the values needed to supply a TRS URL
+        final String basePath = workflowClient.getClient().getGa4Ghv20Api().getApiClient().getBasePath();
+        final String versionId = workflowClient.getVersionID(workflowEntry);
+        final String entryId = workflowClient.getTrsId(path);
+        final String type = "PLAIN_" + workflow.getDescriptorType().getValue();
+
+        // Escape each of the URL path values
+        final String escapedId = client.escapeString(entryId);
+        final String escapedVersionId = client.escapeString(versionId);
+        final String escapedType = client.escapeString(type);
+        final String escapedRelativePath = client.escapeString(workflow.getWorkflowPath());
+
+        // Return the TRS URL for a given entry
+        return MessageFormat.format("{0}/ga4gh/trs/v2/tools/{1}/versions/{2}/{3}/descriptor/{4}",
+            basePath,
+            escapedId,
+            escapedVersionId,
+            escapedType,
+            escapedRelativePath);
+    }
+
+    /**
+     * Attempts to load a file from the provided path. Errors out if the file doesn't exist.
+     *
+     * @param workflowParamPath Path to a file
+     * @return A File object
+     */
+    public static File loadFile(String workflowParamPath) {
+        if (workflowParamPath == null) {
+            return null;
+        }
+
+        // Verify the file path exists
+        Path path = Paths.get(workflowParamPath);
+        if (!Files.exists(path)) {
+            errorMessage(MessageFormat.format("Unable to locate file: {0}", workflowParamPath), CLIENT_ERROR);
+        }
+
+        // TODO: Handle path expansions? i.e. '~/my/file.txt'
+        return new File(workflowParamPath);
+    }
+
+    /**
+     * Given a list of string paths, this attempts to convert it to a list of files.
+     *
+     * @param attachments A list of paths that correspond to files that need to be attached to the WES request
+     * @return A list of File objects
+     */
+    public static List<File> loadAttachments(List<String> attachments) {
```

Comment:
There can be files other than descriptors that are imported by the workflow, e.g. $import and $include in CWL. I don't know whether TRS will retrieve those files along with secondary descriptors.

Reply:
In that case, they'd need to be attached here as well I think. I'm actually not certain how a WES server would decide to switch between resolving by TRS URL and/or using the list of attached files, but seems to be an implementation-specific scenario.

Either way, we should have support for multiple file-upload methods once https://ucsc-cgl.atlassian.net/browse/DOCK-1981 is complete.